addons/web_editor/static/src/js/common/ace-dialog.js

```diff
@@ -0,0 +1,109 @@
+/** @odoo-module **/
+
+import { loadBundle } from "@web/core/assets";
+import Dialog from 'web.Dialog';
+import core, { _t } from 'web.core';
+
+/**
+ * Render an code editor in the given node.
+ * @param {DOMElement} node where the editor will be rendered
+ * @param {String | undefined} content text of the editor
+ * @returns {Object} the ace editor instance
+ */
+export function renderCodeEditor(node, content) {
+    content = content || ''
+    const aceEditor = window.ace.edit(node);
+    aceEditor.setTheme('ace/theme/monokai');
+    aceEditor.setValue(content, 1);
+    aceEditor.setOptions({
+        minLines: 20,
+        maxLines: Infinity,
+        showPrintMargin: false,
+    });
+    aceEditor.renderer.setOptions({
+        highlightGutterLine: true,
+        showInvisibles: true,
+        fontSize: 14,
+    });
+
+    const aceSession = aceEditor.getSession();
+    aceSession.setOptions({
+        mode: "ace/mode/qweb",
+        useWorker: false,
+    });
+    return aceEditor;
+}
+
+/**
+ * Display a dialog to edit the custom code of the website.
+ * @param {OdooWidget} ctx
+ */
+export async function openHeadBodyEditor(widget) {
+    await loadBundle({
+        jsLibs: [
+            '/web/static/lib/ace/ace.js',
+            '/web/static/lib/ace/mode-xml.js',
+            '/web/static/lib/ace/mode-qweb.js',
+        ],
+    });
+
+    let websiteId;
+    widget.trigger_up('context_get', {
+        callback: (ctx) => {
+            websiteId = ctx['website_id'];
+        },
+    });
+
+    let website = (await widget._rpc({
```

Comment:
Instead of an util which receives the parent widget and then calls private method like this here, could not you make a specialized dialog instance which is in charge of doing this RPC when it is being opened? It's fine like this but if we can avoid calling private methods (the ones that start with _), it's better :+1: 

So instead of `openHeadBodyEditor(this)`, I would imagine something like `new HeadBodyDialog(this).open()` and all those libs loading, website read etc would be part of the Dialog lifecycle (willStart, start, open, ...).

Reply:
Great idea I'll look into it :)